### ISSUE TITLE
git lfs migrate import appears to convert files, but doesn't delete original blobs

### ISSUE THREAD

**Issue by agunnerson-ibm** [git-lfs/git-lfs#2984]
We're running into an issue where `git lfs migrate` appears to be converting files to LFS (`.git/lfs` is the expected size), but it seems like the original objects aren't getting deleted.

For example, we'll mirror clone the repo from the original server and push it to the new one.

```
[andrew.gunnerson@<REDACTED> temp]$ git clone --mirror ssh://git@<INTERNAL_GIT_1>:7999/foo/bar.git
Cloning into bare repository 'bar.git'...
remote: Counting objects: 54635, done.
remote: Compressing objects: 100% (16394/16394), done.
remote: Total 54635 (delta 30592), reused 49397 (delta 28181)
Receiving objects: 100% (54635/54635), 260.24 MiB | 34.77 MiB/s, done.
Resolving deltas: 100% (30592/30592), done.

[andrew.gunnerson@<REDACTED> temp]$ cd bar.git/

[andrew.gunnerson@<REDACTED> bar.git]$ git push --mirror git@<INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git
Counting objects: 54635, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (13983/13983), done.
Writing objects: 100% (54635/54635), 260.24 MiB | 43.56 MiB/s, done.
Total 54635 (delta 30592), reused 54635 (delta 30592)
remote: Resolving deltas: 100% (30592/30592), done.
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
remote: error: File a/b/c.jar is 123.23 MB; this exceeds GitHub Enterprise's file size limit of 100.00 MB
To <INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git
 <REDACTED OTHER REFS>
 ! [remote rejected]   master -> master (pre-receive hook declined)
 <REDACTED OTHER REFS>
error: failed to push some refs to 'git@<INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git'
```

The server rejects `a/b/c.jar`, so I move that to LFS. I know the files are successfully moved because the local `lfs` folder in the git dir is non-empty and has the correct size.

```
[andrew.gunnerson@<REDACTED> bar.git]$ git lfs install
Updated git hooks.
Git LFS initialized.

[andrew.gunnerson@<REDACTED> bar.git]$ git lfs migrate import --include='a/b/c.jar' --everything
migrate: Sorting commits: ..., done
migrate: Rewriting commits: 100% (5250/5250), done
  <REDACTED OTHER REFS>
  master                                        babf4668336192046649d5c26f992036e7b64d27 -> 15b87d6d04fb57775c84b8fe2554313989d8514f
  <REDACTED OTHER REFS>
migrate: Updating refs: ..., done

[andrew.gunnerson@<REDACTED> bar.git]$ du -chs lfs/
124M    lfs/
124M    total
```

However, the server still rejects the file.

```
[andrew.gunnerson@<REDACTED> bar.git]$ git push --mirror git@<INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git
Uploading LFS objects: 100% (1/1), 129 MB | 23 MB/s, done
Counting objects: 64440, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (24034/24034), done.
Writing objects: 100% (64440/64440), 261.53 MiB | 25.02 MiB/s, done.
Total 64440 (delta 36819), reused 53957 (delta 30341)
remote: Resolving deltas: 100% (36819/36819), done.
remote: error: GH001: Large files detected. You may want to try Git Large File Storage - https://git-lfs.github.com.
remote: error: File a/b/c.jar is 123.23 MB; this exceeds GitHub Enterprise's file size limit of 100.00 MB
To <INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git
 <REDACTED OTHER REFS>
 ! [remote rejected]   master -> master (pre-receive hook declined)
 <REDACTED OTHER REFS>
error: failed to push some refs to 'git@<INTERNAL_GIT_2>:andrew-gunnerson/bar-test.git'
[andrew.gunnerson@<REDACTED> bar.git]$
```

I suspect that the file might not be deleted after conversion to LFS. Any ideas as to what might be going on? Not sure if it matters, but the repo was originally converted from SVN and the commits that introduced and removed the files were originally SVN commits.

**Comment by ttaylorr** [comment]
Hi @agunnerson-ibm, thanks for opening this issue and I'm sorry that you're having trouble. I think that what might be going wrong is that you have references which _aren't_ being updated by `--everything`, and the `git push --mirror` is trying to push them to your remote, which is causing them to be rejected.

Can you see if there are any references that are listed in `git for-each-ref` that don't appear in the output of `git lfs migrate import`?

If so, I think the best way to go would be either to (a) explicitly migrate them with `--include-ref`, or (b) remove the refs and run `git gc` to remove the orphaned blobs/trees/commits in your repository.

**Comment by agunnerson-ibm** [comment]
> Can you see if there are any references that are listed in git for-each-ref that don't appear in the output of git lfs migrate import?

Thanks! That was exactly it. There was a `refs/pull-requests/27/merge` ref that wasn't included by `--everything`. There's a similarly named `refs/pull-requests/27/from` ref that _was_ included though. Any ideas why this might happen?

**Comment by ttaylorr** [comment]
> Any ideas why this might happen?

Hm, that reference `refs/pull-requests` isn't getting picked up because it doesn't look like a local reference, i..e, that it doesn't start with `refs/heads`. Do you know where that ref was generated from? Either way, I think that loosening this requirement couldn't hurt.

**Comment by agunnerson-ibm** [comment]
Yeah, the `refs/pull-requests` refs are generated by Bitbucket (data center edition). I'm cloning with `--mirror`, so I believe all the refs are available locally. The weird thing is that `git lfs migrate import` recognizes one of these refs, but not the other.

**Comment by ttaylorr** [comment]
Fixed in #3121.
